core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java

```diff
@@ -424,20 +425,33 @@ public void onTopologyUpdate(final CacheTopology cacheTopology, final boolean is
          // Remove the transactions whose originators have left the cache.
          // Need to do it now, after we have applied any transactions from other nodes,
          // and after notifyTransactionDataReceived - otherwise the RollbackCommands would block.
-         if (transactionTable != null) {
-            transactionTable.cleanupLeaverTransactions(rpcManager.getTransport().getMembers());
+         try {
+            if (transactionTable != null) {
+               transactionTable.cleanupLeaverTransactions(rpcManager.getTransport().getMembers());
+            }
+         } catch (Exception e) {
+            // Do not fail state transfer when the cleanup fails. See ISPN-7437 for details.
+            log.transactionCleanupError(e);
          }
 
          commandAckCollector.onMembersChange(newWriteCh.getMembers());
 
+         // The rebalance (READ_OLD_WRITE_ALL) is confirmed through notifyEndOfRebalanceIfNeeded
+         // and STABLE does not have to be confirmed at all
```

Comment:
Yeah, `LocalCacheStatus` would be the place to do it, because it now has the `LimitedExecutorService` to synchronize the execution of the topology updates, and `future.whenComplete()` should be able to replace the executor. But you were probably right to abandon that approach here :)